Fix/circuit breaker

ph opened the pull request:
This pull request aim to applies the pipeline back pressure up to logstash-forwarder (LSF)  clients.
In the current implementation the lumberjack input create a new thread per LSF connection and start accepting new data right away, if the **pipeline** is slow or there is a transient or a permanent error on the output, each connection threads will block and remain in this state until the pipeline unblock. On the LSF side theses block will be detected as a connection timeout and they will reconnect to logstash and restransmit the event frame, Logstash will accept this new connection, read the data and block again. LSF will never stop restransmitting data and logstash will create threads until the process
is out of memory.

This pull request bring a few features to the lumberjack input:

* A Small buffered queue with a blocking timeout, this value is a bit higher than the LSF threshold, a connection thread will never stay open forever.
* A Circuit breaker, if there is too many timeout errors the breaker will open and logstash will start refusing new connection for a period of time.
* Threads are now added to a cached threadpool backed by concurrent-ruby and java ThreadPoolExecutor (currently unbounded)

*The timeout option of the size queue will be migrated to the main logstash queue in a future pull request*

## RUN 
**Platform**:
MacBook Pro (Retina, 13-inch, Mid 2014) (no completly idle so it may impact results)
MacOS X 10.10.3
3 GHz Intel Core i7
16 GB 1600 MHz DDR3

**Config**
Logstash: https://gist.github.com/f0a2ef9566c4cd35c6bf
Logstash-Forwarder: https://gist.github.com/a55d409c68c18cd48c28

Logstash command: bin/logstash -f ~/es/logstash-config/lumberjack_test/input.conf | pv -Wbart > /dev/null

Sample.log  ~ 5.1 gigs file (Let the tests ran for ~4, enough to warm the jvm)
Extract of the log: https://gist.github.com/061c05fbade8c4eddd52

## Performance tests

| connection | Baseline v0.1.5 | SizeQueue + Timeout + Circuit Breaker |
| --------------- | -------------------- | ---------------------------------------------------- |
| 1  | 5.71MiB 0:04:00 [28.3KiB/s] [24.4KiB/s] | 6.61MiB 0:04:07 [27.6KiB/s] [27.4KiB/s] |
| 3  | 9.88MiB 0:04:02 [37.2KiB/s] [41.8KiB/s] | 9.13MiB 0:04:04 [38.8KiB/s] [38.3KiB/s] |

##### Resiliency tests

Config:

Logstash: https://gist.github.com/f0a2ef9566c4cd35c6bf
Logstash-Forwarder: https://gist.github.com/a55d409c68c18cd48c28

Scenario:
Permanent error

Setup: 1, LSF -> Logstash -> ES
Stop ES, LSF timeout, try to reconnect and get connection refused
Restart ES, LSF reconnect and resend events

Using the sleep filter we can simulate congestion in the pipeline, if there is too much congestion LSF will timeout and reconnect. The threads will be killed by timeout event and free the memory or they will

ph commented:
@suyograo @jordansissel can you look at this?

suyograo commented:
To clarify: This is not being targeted for LS 1.5

jordansissel commented:
I tested this manually with:

```
# Logstash
bin/logstash --debug -e 'input { lumberjack { port => 4444 ssl_key => "/users/jls/projects/logstash-forwarder/lumberjack.key" ssl_certificate => "/users/jls/projects/logstash-forwarder/lumberjack.crt" } } filter { sleep { time => 1 } }'

# Logstash-forwarder
seq 10 | xargs -P10 -n1 sh -c 'yes "hello world $1" | ./logstash-forwarder -config test.conf' -
```

10 concurrent logstash-forwarders shipping logs to logstash. Looks good so far.

I was able to observe the new behaviors:
* Termination of existing connections if the pipeline is stalled
* Ignorance of new connections.

On the logstash-forwarder side, these behaviors manifested as different kinds of faults:

* `Read error looking for ack: EOF` - This was an indication that the connection was terminated by Logstash (expected, assuming the OpenBreaker exception is thrown inside the Lumberjack Connection thread, causing a connection closure)

* `Failed to tls handshake with ::1 read tcp [::1]:4444: i/o timeout` - This was an indication that the TCP connection was accepted by the kernel, but was never accepted by Logstash, so the TLS handshake simply timed out. (expected congestion control behavior)

@ph does my observation match your expected behavior?

ph commented:
@jordansissel your observation is correct on with the behavior a I have implemented. 
Did you try with a recovery scenario?

We have to keep in mind that it could produce duplicate events in some case, but I think it follow our contract of a `least once`

ph commented:
By recovery scenario, I mean recovery from a permanent failure like ES went down.

jordansissel commented on lib/logstash/circuit_breaker.rb:
```diff
@@ -0,0 +1,88 @@
+    DEFAULT_TIME_BEFORE_RETRY = 30 # LSF reconnect value is 15 second
+    DEFAULT_EXCEPTION_RESCUED = [StandardError]
+
+    def initialize(name, options = {}, &block)
```
Can you document the options?

jordansissel commented on lib/logstash/inputs/lumberjack.rb:
```diff
@@ -31,47 +31,77 @@ class LogStash::Inputs::Lumberjack < LogStash::Inputs::Base
-            output_queue << event
+      begin
+        # Wrapping the accept call into a CircuitBreaker
+        if @circuit_breaker.close?
```
feels like this should be `.closed?` (closed, not close) when I read it

jordansissel commented on lib/logstash/inputs/lumberjack.rb:
```diff
@@ -31,47 +31,77 @@ class LogStash::Inputs::Lumberjack < LogStash::Inputs::Base
+            end
           end
+        else
+          @logger.warn("Lumberjack input: the pipeline is blocked, temporary refusing new connection.")
```
I wonder if we should put more details here, or link to a "read more" webpage about what is going on internally? 

My worry is that this log message will just cause users to create bug reports with the log message, not knowing that the log message indicates that the pipeline is stuck further down (in a filter or an output)

ph replied:
Agree, what do you think If I create a document about back pressure?
We have seen it people people will just create new issues or hack values.

jordansissel replied:
+1 on documenting it, at least then we can redirect users to the doc to hopefully clarify things.

jordansissel commented on lib/logstash/inputs/lumberjack.rb:
```diff
@@ -31,47 +31,77 @@ class LogStash::Inputs::Lumberjack < LogStash::Inputs::Base
+          @logger.warn("Lumberjack input: the pipeline is blocked, temporary refusing new connection.")
+          sleep(RECONNECT_BACKOFF_SLEEP)
         end
+      rescue LogStash::CircuitBreaker::OpenBreaker => e
```
Can you comment within the rescue block (for future devs) why it's safe to ignore this?

jordansissel commented on lib/logstash/sized_queue_timeout.rb:
```diff
@@ -0,0 +1,59 @@
+      @timeout = options.fetch(:timeout, DEFAULT_TIMEOUT)
+    end
+
+    def push(obj)
```
Should push take a timeout argument:

```ruby
def push(obj, timeout=DEFAULT_TIMEOUT)
   ...
```

Shave an instance variable for something that maybe should be a decision made at push() time, not initialization? Thoughts?

ph replied:
+ 1 once we get the logic merge in the @colinsurprenant's persistent queue plugin developers could use their own tolerance values.

jordansissel commented on spec/logstash/size_queue_timeout_spec.rb:
```diff
@@ -0,0 +1,65 @@
+require "logstash/sized_queue_timeout"
+
+describe "LogStash::SizedQueueTimeout" do
+  let(:max_size) { 3 }
```
If you want to add randomization here, do this:

```ruby
require "flores/random"

...
  let(:max_size) { Flores::Random.integer(2..100) }
  let(:element) { Flores::Random.text(0..100) }
```

jordansissel commented on spec/support/logstash_test.rb:
```diff
@@ -0,0 +1,33 @@
+    attr_reader :ssl_key, :ssl_cert
+
+    def initialize
+      @ssl_cert = Stud::Temporary.pathname("ssl_certificate")
```
No change required, just wanted to let you know: I added some PKI support to Flores, so soon we'll be able to generate SSL certificates from Ruby directly for easier testing.

jordansissel commented on spec/inputs/lumberjack_spec.rb:
```diff
@@ -4,54 +4,46 @@
-
+  let(:connection) { double("connection") }
+  let(:certificate) { LogStashTest.certificate }
+  let(:port) { LogStashTest.random_port }
```
Longer term, we can use Flores for this, Flores::Random.tcp_listener will get you a socket listening on a random port, and we can use rspec mocks to inject this tcp socket into the Lumberjack Server. :)

ph commented:
@jordansissel I have adressed most of the comments here. Created https://github.com/logstash-plugins/logstash-input-lumberjack/issues/28 to address the documentation issue.

I used flores for the random values when it support the certificate I'll update the test :dancer:

jordansissel commented on spec/support/logstash_test.rb:
```diff
@@ -21,3 +21,13 @@ def random_port
+  end
+  #
+  # wait for the thread to spin
+  sleep(0.1)
```
If you explicitly want to yield, it's probably best to use a ConditionVariable here and have the thread wake you up when it's started.

jordansissel replied:
Although I'd prefer not needing the wait; what's the goal?

ph replied:
its only a test helper to make sure the thread is ready before doing any tests on it.

jordansissel commented on spec/logstash/size_queue_timeout_spec.rb:
```diff
@@ -0,0 +1,68 @@
+        subject << element
+      end
+
+      expect(blocked.status).to eq("sleep")
```
maybe instead of the sleep in `testing_thread`, we could use Stud::Try?

```ruby
Stud::try(5.times) do
  expect(blocked.status).to eq("sleep")
end
```

The above will do exponential backoff (sleeping) up to 5 times while the block (`expect ....`) is failing (exception()

ph replied:
The thread is not yet blocked when the expectation is run, so the 

```ruby
      blocked = Thread.new do
        subject.pop_no_timeout
      end

      Stud::try(5.times) do
       puts "First"
        expect(blocked.status).to eq("sleep")
      end
```

You only get one "First" message and this raise.
```
  1) LogStash::SizedQueueTimeout when the queue is empty block on pop
     Failure/Error: expect(blocked.status).to eq("sleep")

       expected: "sleep"
            got: "run"

       (compared using ==)
     # ./spec/logstash/size_queue_timeout_spec.rb:59:in `(root)'
     # ./spec/logstash/size_queue_timeout_spec.rb:58:in `(root)'
```

jordansissel replied:
Weird, this means taht rspec's failure exception doesn't inherit from StandardError, so `Stud::Try` never catches it (https://github.com/jordansissel/ruby-stud/blob/master/lib/stud/try.rb#L95)

I can update Stud::Try to catch all exceptions instead of just the family of StandardError. Thoughts?

If I do that, then the Stud::try(5.times) will fail a few times, sleep each time (exponential backoff) hoping that the next attempt will succeed and not raise an error.

jordansissel replied:
Confirmed rspec's expectation failure is not a descendent of StandardError: 

```
#<RSpec::Expectations::ExpectationNotMetError: expected: == 5
     got:    0>

>> RSpec::Expectations::ExpectationNotMetError.ancestors
=> [RSpec::Expectations::ExpectationNotMetError, Exception, Object, PP::ObjectMixin, Kernel, BasicObject]
```

ph replied:
Would it be better to create a specific rspec helper for this if scenario?

````
retry_until_thread_ready {  expect ... }
```

ph replied:
I think it would hide the complexity and we can have a baseline number for theses kind of retries.

driskell replied:
Ruby own way of testing queues is interesting. Don't know if might be a reference point?

See `grind`:
https://github.com/ruby/ruby/blob/trunk/test/thread/test_queue.rb

ph replied:
https://github.com/ruby/ruby/blob/trunk/test/thread/test_queue.rb#L122
A bit similar to what we do on the pipeline testing.

ph replied:
@driskell Honestly I think I prefer the way `test_queue` is doing it, I'll discuss with @jordansissel about it.

jordansissel commented on spec/logstash/size_queue_timeout_spec.rb:
```diff
@@ -0,0 +1,68 @@
+    end
+  end
+
+  context "when the queue is empty" do
```
There's test cases for fully and empty, but not occupied-but-not-full. Can you add tests for that? You can use flores for this:

```ruby
Flores::Random.iterations(0..max_size) do 
  subject << "hurray"
end
```

^^^ should only populate a partially, but not fully.

ph replied:
+1

ph commented:
@jordansissel Even if I have added the `Try` rspec helper (I see the need in other integration tests), I believe that using the `sleep(0.1)` as a way to synchronize when the threads are ready to be tested is the way to do it. I have changed a few things to lower the sleep at a minimum. Can you review the changes I've made?

@driskell's reference to the ruby's sizequeue made me consider this.

jordansissel commented:
LGTM

ayashjorden commented:
Hi,
In 'lib/logstash/inputs/lumberjack.rb' there's a TODO comment about SSL CA authentication.
does that mean that this PR won't authenticate clients?

We do suffer from the subject of this PR and would like start experimenting with this plugin update.

Thanks,
Yarden

ayashjorden commented:
Hi,
Any documentation for the Circuit breaker ?
